recipe/patches/0005-CMake-Explicitly-link-with-CCCL-standalone-or-CTK-10.patch

```diff
@@ -0,0 +1,60 @@
+From 15b7012f54793f3cf807501453a61b6a6c1fb3c3 Mon Sep 17 00:00:00 2001
+From: Philip Hyunsu Cho <chohyu01@cs.washington.edu>
+Date: Tue, 23 Jul 2024 18:42:54 -0700
+Subject: [PATCH] [CMake] Explicitly link with CCCL (standalone or CTK)
+ (#10624)
+
+---
+ CMakeLists.txt    | 18 ++++++++++++++++++
+ cmake/Utils.cmake |  6 +++---
+ 2 files changed, 21 insertions(+), 3 deletions(-)
+
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index dc0f7e6..aa07840 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -224,6 +224,24 @@ if(USE_CUDA)
+   add_subdirectory(${PROJECT_SOURCE_DIR}/gputreeshap)
+ 
+   find_package(CUDAToolkit REQUIRED)
++  find_package(CCCL CONFIG)
++  if(NOT CCCL_FOUND)
++    message(STATUS "Standalone CCCL not found. Attempting to use CCCL from CUDA Toolkit...")
++    find_package(CCCL CONFIG
++      HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++    if(NOT CCCL_FOUND)
++      message(STATUS "Could not locate CCCL from CUDA Toolkit. Using Thrust and CUB from CUDA Toolkit...")
++      find_package(libcudacxx CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(CUB CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(Thrust CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      thrust_create_target(Thrust HOST CPP DEVICE CUDA)
++      add_library(CCCL::CCCL INTERFACE IMPORTED GLOBAL)
++      target_link_libraries(CCCL::CCCL INTERFACE libcudacxx::libcudacxx CUB::CUB Thrust)
++    endif()
++  endif()
+ endif()
+ 
+ if(FORCE_COLORED_OUTPUT AND (CMAKE_GENERATOR STREQUAL "Ninja") AND
+diff --git a/cmake/Utils.cmake b/cmake/Utils.cmake
+index 8659df7..ec69278 100644
+--- a/cmake/Utils.cmake
++++ b/cmake/Utils.cmake
+@@ -108,12 +108,12 @@ function(xgboost_set_cuda_flags target)
+     target_compile_definitions(${target} PRIVATE -DXGBOOST_USE_NVTX=1)
+   endif()
+ 
++  target_link_libraries(${target} INTERFACE CCCL::CCCL)
++  target_link_libraries(${target} PUBLIC CUDA::cudart_static)
```

Comment:
In terms of getting the include order right, this is doing what we want as [seen on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=986631&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=1023 ) (as opposed to the reversed order seen previously)

```bash
-I$PREFIX/include ... -isystem /usr/local/cuda/targets/x86_64-linux/include
```

However am seeing some odd errors

```cpp
$PREFIX/include/cuda/std/__utility/pair.h(560): error: no operator "==" matches these operands
```

So likely there is still something off here

Reply:
Fix available at https://github.com/dmlc/xgboost/pull/10633

Reply:
Updated the patch.

Reply:
It turns out the particular error
```
$PREFIX/include/cuda/std/__utility/pair.h(560): error: no operator "==" matches these operands
```
was due to a bug in CCCL. So we should probably set a filter on the CCCL version to exclude the affected versions (2.4.0, 2.5.0).

Reply:
Ah good catch. Thanks Hyunsu! 🙏

So we were picking up the newer CCCL. We just ran into a known bug

Here's hoping we are in the clear 🍀

Reply:
Yep, we are now using the latest CCCL for sure.